### ISSUE TITLE
memory leak

### ISSUE THREAD

**Issue by Seohanul** [cocos2d/creator_to_cocos2dx#150]
when I put this code in button callback, It cause memory leak.

for (int i = 0; i < 10; ++i)
{
			creator::CreatorReader * creatorReader = creator::CreatorReader::createWithFilename ("filename");

			creatorReader->setup ();

			/*creatorReader->autorelease ();*/
}

every time, momory goes up. 
how can I fix this problem?

**Comment by Seohanul** [comment]
I enabled CC_REF_LEAK_DETECTION, and I did a compare with printleaks before and after loading the creatorReader, I find there are some SpriteFrameCache and SpriteFrame which are not released

**Comment by Seohanul** [comment]
I did a Director::getInstance ()->purgeCachedData (); and the spriteframes are not being released, I see their ref count is 2 when purgecacheddata is called

**Comment by Seohanul** [comment]
Reporting the first leak (and fix),

When creating a button, there is a child label which is created as a protectedchild node using in Ln400 of CreatorReader.cpp (button->setTitleLabel), the created Label is not set to autorelease so it will leak.

**Comment by drelaptop** [comment]
Could you supply a test case to show the memory leak? for example to modify the default cpp-empty-test, add your code, and push it into GitHub, tell me the link. 

And I didn't think `Ln400 of CreatorReader.cpp (button->setTitleLabel)` cause memory leak. the Label added into button is created by 
```
case buffers::AnyNode_Label:
            node = createLabel(static_cast<const buffers::Label*>(buffer));
            break;
```
So it is setted into autorelease pool

**Comment by Seohanul** [comment]
![label memory leak](https://user-images.githubusercontent.com/31907366/37806307-3b2f86fe-2e7a-11e8-9788-5babe54b2909.jpg)

If label is child of button, CreatorReader.cpp (button->setTitleLabel) set a leabel to protected child.
I'm not sure this can be released properly.
I just set label->autorelease() after button->setTitleLabel, and no more leak.

**Comment by Seohanul** [comment]
And there are other issue. 
AnimationManager not to be relased for looping animations.

**Comment by Seohanul** [comment]
![animation memory leak](https://user-images.githubusercontent.com/31907366/37807186-11f97f7e-2e7f-11e8-9a6a-8fa8d0b4f3d1.jpg)
And there are other issue. 
AnimationManager not to be relased for looping animations.
I play some animations at the same time as loading creator file. and looping.
And after release all nodes, animation still remaining.
ActionManager doesn't have fuction for stop 'all' animation or remove 'all' animation clip.
And also I think looping animation is not setted into autorelease pool.
How can I release looping animations properly?

**Comment by Seohanul** [comment]
Thank you for reply.

I have another question about animation manager.

![virtual dtor](https://user-images.githubusercontent.com/31907366/37884734-07560b66-30e4-11e8-99ed-e10368d08691.jpg)

Animation Manager inherited by cocos2d::Node.

So, I think destructor of Animation Manager should be virtual. Am I right?

**Comment by drelaptop** [comment]
I think so too, but it's not a matter thing for not any sub-class inherit `AnimationManager`.

This comment support __Markdown__, so you can add a code block by markdown, not need to paste a picture for a code block.

**Comment by imtrobin** [comment]
Without the virtual dtor, it did not call the Node destructor properly so there was a leak.

**Comment by drelaptop** [comment]
Node's destructor is virtual, it do call the Node destructor when `AnimationManager ` destruct. I'm afraid that you are misunderstanding this.

**Comment by drelaptop** [comment]
@Seohanul see commit [fix loop animate memory leak](https://github.com/cocos2d/creator_to_cocos2dx/pull/152/commits/449c242ed89a7e0fc1fbb1c78d5e313ee4945cea) ,[fix PlayOnLoad loop animation leaks](https://github.com/cocos2d/creator_to_cocos2dx/pull/152/commits/ba726bc5f1880f3234e2370b8150cf4662ab32c3) in PR #152 to fix the memory leak

if existing a loop animation, please stop it manually, using function:
```C++
    // if AnimationClip is stopped, can not run it again.
    void stopAnimationClip(cocos2d::Node *target, const std::string &animationClipName);
    // if a "Play On Load" animation is a loop animation, please stop it manually.
    void stopPlayOnLoad();
```
`stopAnimationClip` stop the animation started by `playAnimationClip`, `stopPlayOnLoad` stop the animation started by `playOnLoad` automatically.

**Comment by imtrobin** [comment]
Why don't you call this stop on node dtor automatically? If its playOnLoad automatically,  it is done so in creator studio, it should be able to unload automatically.

**Comment by drelaptop** [comment]
Indeed, it's reasonable, but hard to do it in this plugin, I tried. Strictly `playOnLoad` isn't automatically, it's called by the `CreatorReader::getSceneGraph` function,  `stopPlayOnLoad` should be put into what place, `CreatorReader` `AnimationManager` destruction function or somewhere else?

suggest you merge that PR, and have a try.

**Comment by nunu-e64** [comment]
How about `stopPlayOnLoad` called in `AnimationManager` destructor and stop automatically?
If animations played on load has already stopped, `stopPlayOnLoad` do nothing because cashed animation would be not found.
https://github.com/cocos2d/creator_to_cocos2dx/blob/master/creator_project/packages/creator-luacpp-support/reader/animation/AnimationManager.cpp#L65

**Comment by nunu-e64** [comment]
Sorry, its a bad idea.
`Animation Manager` is retained when running animation clip. So `Animation Manager` will not be destructed.
https://github.com/cocos2d/creator_to_cocos2dx/blob/master/creator_project/packages/creator-luacpp-support/reader/animation/AnimationManager.cpp#L91

**Comment by drelaptop** [comment]
thanks for your review, I hold the same idea with you.
